sst/sstable.go

```diff
@@ -161,18 +169,30 @@ func BuildSSTable(format common.DataFormat, buffSizeEstimate int, entriesEstimat
 		indexOffset:      uint32(indexOffset),
 		creationTime:     uint64(time.Now().UTC().UnixMilli()),
 		data:             buff,
-	}, smallestKey, largestKey, minVersion, maxVersion, nil
+	}
+
+	selfTable.metadataBinary = selfTable.serializedMetdata()
+	selfTable.metadataSize = len(selfTable.metadataBinary)
+
+	return selfTable, smallestKey, largestKey, minVersion, maxVersion, nil
+}
+
+func (s *SSTable) serializedMetdata() []byte {
+	buff := make([]byte, 0, maxMetadataSize)
+	buff = binary.AppendUvarint(buff, uint64(s.maxKeyLength))
+	buff = binary.AppendUvarint(buff, uint64(s.numEntries))
+	buff = binary.AppendUvarint(buff, uint64(s.numDeletes))
+	buff = binary.AppendUvarint(buff, uint64(s.numPrefixDeletes))
+	buff = binary.AppendUvarint(buff, uint64(s.indexOffset))
+	buff = binary.AppendUvarint(buff, s.creationTime)
+
+	return buff
 }
 
 func (s *SSTable) Serialize() []byte {
 	// To avoid copying the data buffer, we put all the meta-data at the end
-	buff := encoding.AppendUint32ToBufferLE(s.data, s.maxKeyLength)
-	buff = encoding.AppendUint32ToBufferLE(buff, s.numEntries)
-	buff = encoding.AppendUint32ToBufferLE(buff, s.numDeletes)
-	buff = encoding.AppendUint32ToBufferLE(buff, s.numPrefixDeletes)
-	buff = encoding.AppendUint32ToBufferLE(buff, s.indexOffset)
-	buff = encoding.AppendUint64ToBufferLE(buff, s.creationTime)
-	return buff
+
+	return append(s.data, s.metadataBinary...)
```

Comment:
Simpler and more efficient to just do the varint appending directly in here.

Reply:
But I will need the size of metadata buffer ahead of time during `BuildSSTable` as it's used in `SizeBytes`. Hence the need of the function `serializedMetdata` . Any suggestion around the `SizeBytes` while avoiding buffer copy ?

Reply:
If you want to know the size of the metadata that you can just measure len() before and after appending it to the buffer, no?

Reply:
Yes and the new function I am about to commit does that. but that assumes the SizeByte is only called after a call to `Serialize` . Or are you suggesting to just append the metadata when creating the `SSTable struct instance`.

Reply:
With your changes you're already creating the serialized metadata in the BuildSSTable method. But you're building that into a new buffer which you then append to the main buffer on Serialize. That seems unnecessary. Could you not just append it into the same buffer with all the rest of the data?

Reply:
Yes I did that.